mlir/lib/IR/AsmPrinter.cpp

```diff
@@ -1630,6 +1648,23 @@ void SSANameState::numberValuesInOp(Operation &op) {
       asmInterface.getAsmBlockNames(setBlockNameFn);
       asmInterface.getAsmResultNames(setResultNameFn);
     }
+    if (!opAsmOpInterfaceUsed) {
+      // If the OpAsmOpInterface didn't set a name, and
+      // all results have OpAsmTypeInterface, get names from types.
+      bool allHaveOpAsmTypeInterface =
+          llvm::all_of(op.getResultTypes(), [&](Type type) {
+            return mlir::isa<OpAsmTypeInterface>(type);
+          });
```

Comment:
Why check if they all implement the interface?

Reply:
For an op with multiple results, we could not suggest a meaningful result "group" name. 

For example, in the following assembly the second result should not bear the name of `type_name`. It is the responsibility of `OpAsmOpInterface` to handle the grouping behavior.

```mlir
%type_name:2 = test.some_op %arg : i32 -> (!test.type, i32)
```

We should rather suggest the following way to clearly separate them.

```mlir
%type_name, %0 = test.some_op %arg : i32 -> (!test.type, i32)
```

But the default grouping behavior is not this way, so I only apply such separation when all of the results have `OpAsmTypeInterface`.